text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,149 @@
+
+> @GuillaumeGomez comment: This is the "story" part. Is it ok to put it here or should it be somewhere else? Also: is it clear enough?
+
+A good example where this would be very helpful is for the [rust compiler source code](https://doc.rust-lang.org/1.57.0/nightly-rustc/src/rustc_middle/hir/map/mod.rs.html#872-877): even if a fully set up IDE or github, it's very hard to go around without previous knowledge or help from someone else. It's also very common to have undocumented items or to just want to see how something is implemented. With the "jump to definition", it already improved quite a lot the browsing experience by allowing to jump super quickly by simply clicking on a item. The missing part is now to go back from the source code into the item documentation to make it complete.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When going to the source code pages (as you can see already in the [compiler documentation](https://doc.rust-lang.org/1.57.0/nightly-rustc/src/rustc_middle/hir/map/mod.rs.html#872-877)), you will be able to click on idents to go to their definition.
+
+Once at their definition, if you click on the item name, you'll arrive on its documentation page (if it has one, otherwise nothing will happen).
+
+An interesting thing to note: primitive types always link to documentation. Also, to make it more obvious if a link points to source code or to documentation, they should have different visual markers (different background/underline colors for example).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+It doesn't require javascript since it's only generating links, making it work on both desktop and mobile, with or without javascript enabled.
+
+So on the technical side, there is no big issues. You can see the first part of this RFC implemented [here](https://github.com/rust-lang/rust/pull/84176).
+
+For the UI, we have the following constraints:
+
+ * Must work on mobile and desktop: we cannot rely on mouse hover events.
+ * It's better if it works without javascript (it's important for accessibility, but also for maintenance reasons too: less JS code, less maintenance).
+
+With this in mind, the suggested UI would be like this:
+
+ * Links generated on idents in source code will jump to the ident definition.
+ * Links generated on items' definition will jump to their documentation page (if any, private items don't have one by default).
+ * On mobile, these links will have dotted underline (because you don't have a cursor which change when hovering them).
+
+Basically, the source code pages' UI doesn't change.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+**The biggest concern is actually about "scope expansion"**. If we add this feature, why not adding others like a "commit source code viewer", or a "git blame source code viewer", or "go to this line in github/gitlab"?
```

Comment:
This should also mention the most likely second feature: "find references." That is something I'd really like, and I think many other people would like too. We even got a request related to that [in Zulip yesterday](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/indirect.20type.20mentions).

Reply:
I don't consider it as a drawback but more like a potential extension.

Reply:
I agree - I don't consider it a drawback. But given that (a) some of the team are worried about scope expansion, and (b) it is a scope expansion, you should address it somewhere in the doc, for instance to say "in the future, we are likely to implement 'find references' as well. Here's why that will fit well with the jump to definition feature we are implementing now"